### PR TITLE
#3 Add user creation to the Bulk API end-point

### PR DESCRIPTION
**Pull request by jon-betts**

### CODE REVIEW

File: tests/h/services/bulk_executor/_actions_test.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+from h_matchers import Any
+
+from h.h_api.bulk_api import Report
+from h.h_api.exceptions import ConflictingDataError, UnsupportedOperationError
+from h.models import User
+from h.services.bulk_executor._actions import UserUpsertAction
+from tests.h.services.bulk_executor.conftest import CommandFactory
+
+
+def assert_models_match_data(models_by_name, attrs_by_name, handlers=None):
```
Comment (seanh) [inline tests/h/services/bulk_executor/_actions_test.py:13]: `handlers` seems to be unused.

If there's anything that could be done to simplify `assert_models_match_data()` and `assert_users_match_commands()`, making them "dumber", I think that'd probably be a good idea. As they stand I think they're a bit on the complicated side for test code. They're hard to follow, and look like they need unit tests of their own.

If the tests could be easier to understand then it'd be easier to look at them and verify that they correctly test that the SQLAlchemy code is correct.

I think it'd be okay for each test to know what models it expects to find in the DB and to express those expected models explicitly, even if that results in some duplication in the tests.

Each test could do something along the lines of:

```python
assert db_session.query(models.User) == [
    UserModelWith(username="user_1", ...),
    UserModelWith(username="user_2", ...),
    UserModelWith(username="user_3", ...),
]
```

(where `UserModelWith()` is a simple/classic-style matcher-class, i.e. a class with an `__eq__()`, just defined in this test module)

File: tests/h/services/bulk_executor/_actions_test.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+        action = UserUpsertAction(db_session)
+        action.execute(commands)
+        reports = action.execute(update_commands)
+
+        assert reports == Any.iterable.comprised_of(Any.instance_of(Report)).of_size(3)
+
+        self.assert_users_match_commands(db_session, update_commands)
+
+    def test_if_tails_with_duplicate_identities(self, db_session):
```
Comment (seanh) [inline tests/h/services/bulk_executor/_actions_test.py:50]:
```suggestion
    def test_it_fails_with_duplicate_identities(self, db_session):
```

File: tests/h/services/bulk_executor/_actions_test.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+    @pytest.mark.parametrize("field", ["username", "authority"])
+    def test_it_fails_with_mismatched_queries(self, db_session, field):
+        command = CommandFactory.user_upsert(extras={field: "value"})
+        command.body.query[field] = "DIFFERENT"
+
+        with pytest.raises(UnsupportedOperationError):
+            UserUpsertAction(db_session).execute([command])
+
+    def test_if_fails_with_unsupported_queries(self, db_session):
```
Comment (seanh) [inline tests/h/services/bulk_executor/_actions_test.py:81]:
```suggestion
    def test_it_fails_with_unsupported_queries(self, db_session):
```

File: tests/h/services/bulk_executor/_actions_test.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+        assert reversed_ids == list(reversed(ids))
+
+    @pytest.mark.parametrize("field", ["username", "authority"])
+    def test_it_fails_with_mismatched_queries(self, db_session, field):
+        command = CommandFactory.user_upsert(extras={field: "value"})
+        command.body.query[field] = "DIFFERENT"
+
+        with pytest.raises(UnsupportedOperationError):
+            UserUpsertAction(db_session).execute([command])
```
Comment (seanh) [inline tests/h/services/bulk_executor/_actions_test.py:79]: You should test the exception's message using `match=`. This could be failing because of a disallowed query key rather than because of a query/body-attributes mismatch, and this test would still be passing.

File: h/services/bulk_executor/_actions.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        ]
+
+    def _upsert_user_table(self, users):
+        stmt = self._upsert_statement(
+            User,
+            users,
+            index=[
+                # The text wrapper here prevents SQLAlchemy from quoting the
+                # string, and we need it to be verbatim
+                text("lower(replace(username, '.'::text, ''::text)), authority")
```
[the remark this answers is not in the document]
Comment (jon-betts) [reply h/services/bulk_executor/_actions.py:105]: I don't know, we need what this serialises as in SQL. We could try calling `text(_normalise_username())` and see if that does what we expect?

File: tests/h/services/bulk_executor/_actions_test.py (hunk 1 above)
Comment (jon-betts) [reply tests/h/services/bulk_executor/_actions_test.py:13]: I've been after something like this for a while, I'll see if I can come up with an extension for `Any` that does something like this.

File: h/services/bulk_executor/_actions.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        raise NotImplementedError()
+
+    @staticmethod
+    def _check_upsert_queries(batch, expected_keys):
+        """Check that the query matches expectations and the attributes.
+
+        :param batch: A collection of command objects
+        :param expected_keys: The keys which must be in the query and match
+                              the attributes with the same name.
+        """
```
[the remark this answers is not in the document]
Comment (jon-betts) [reply h/services/bulk_executor/_actions.py:37]: I'll have a go at reworking it. There's no fundamental reason you can't (or we won't one day want to) search by one thing and modify another in theory, but there's a couple of reasons why this doesn't work right now:

* We aren't really searching them update or creating, we are creating first, then updating if there's a conflict
* So we rely on the create conflicting with the existing record we want to update
* If the fields are different, it won't conflict, so this method wouldn't work

I'm not sure what would be involved in a general solution (and we definitely don't need it right now), but we probably couldn't get away with the "ON CONFLICT" to optimise it the way we have here. Hopefully it won't come up!

[the remark this answers is not in the document]
Comment (jon-betts) [reply h/services/bulk_executor/_actions.py:37]: I've updated the description liberally stealing from yours. Thanks!

File: h/services/bulk_executor/_actions.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jon-betts) [reply h/services/bulk_executor/_actions.py:105]: I've tried this a few different ways and I can't get anything to work. Basically we shouldn't be putting functions in our indexes if we want anything like this to be pleasant, as it prevents us from inferring the index by column names (which is the work around to the work around of not being able to specify the index by name, because it isn't a "constraint" even though internally to Postgres it is...)

The advice I've seen is to have an additional normalised field derived from the main field with the function, and then make the compound index from that field without the function. Then we could infer the index by specifying the two fields by name.

File: tests/h/services/bulk_executor/_actions_test.py (hunk 4 above)
Comment (jon-betts) [reply tests/h/services/bulk_executor/_actions_test.py:134]: I think I'm going to stick with this for a few reasons:

 * The interface is that we raise this error when this condition arises, as long as we do, then this test should pass
 * The string of the error shouldn't be part of the interface
 * If we are concerned that this code doesn't explore all the code paths, then I think coverage has us... _covered_ (I'm here all night)

The last thing is to see the tests as boxing in the problem from different sides. The happy path sets a boundary proving we don't raise this exception every time. We then modify one thing which creates the situation we need to detect.

There are some things I think in this situation we can do:

 * Add a second parameter to the exception which allows them to be distinguished
 * Add more specific error classes to allow us to distinguish the cases by class

In this case, as the whole thing is really belt and braces (it should never happen with `h_api` mediating the communication) an assert would probably be enough, so I don't think it's justified.

File: h/services/bulk_executor/_actions.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (seanh) [reply h/services/bulk_executor/_actions.py:118]: Fair enough
